File: Assets/Scripts/GoalController.js
#pragma strict

private var levelGenerator : LevelGenerator;
private var loadNextLevelAt : float;

function Start () {
	levelGenerator = GameObject.Find("LevelGenerator").GetComponent(LevelGenerator);
	loadNextLevelAt = 0.0;
}

function Update () {
	if (loadNextLevelAt && Time.time >= loadNextLevelAt)
		levelGenerator.StartNextLevel();
}

function OnTriggerEnter (other : Collider) {
	if (other.CompareTag("Player")) Reached();
}

function Reached () {
	loadNextLevelAt = Time.time + levelGenerator.LevelParameters().nextLevelDelay;
	transform.animation.Play();
	transform.audio.Play();
	SetPlayerVictorious();
	SetMonstersBeaten();
}

function SetPlayerVictorious () {
	GameObject.FindGameObjectWithTag("Player").
		GetComponent(PlayerController).
		SetVictorious();
}

function SetMonstersBeaten () {
	for (var m : GameObject in GameObject.FindGameObjectsWithTag("Monster")) {
		m.GetComponent(MonsterController).SetBeaten(true);
	}
}

File: Assets/Scripts/GoalController.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class GoalController : MonoBehaviour
{
    private LevelGenerator levelGenerator;
    private float loadNextLevelAt;
    private bool goalReached = false;

    void Start()
    {
        levelGenerator = GameObject.Find("LevelGenerator").GetComponent<LevelGenerator>();
        loadNextLevelAt = 0.0f;
    }

    void Update()
    {
        if (goalReached && Time.time >= loadNextLevelAt)
        {
            levelGenerator.StartNextLevel();
        }
    }

    void OnTriggerEnter(Collider other)
    {
        if (other.CompareTag("Player")) Reached();
    }

    private void Reached()
    {
        goalReached = true;
        loadNextLevelAt = Time.time + levelGenerator.LevelParameters().nextLevelDelay;
        transform.GetComponent<Animation>().Play();
        transform.GetComponent<AudioSource>().Play();
        SetPlayerVictorious();
        SetMonstersBeaten();
    }

    private void SetPlayerVictorious()
    {
        GameObject.FindGameObjectWithTag("Player").
            GetComponent<PlayerController>().
            SetVictorious();
    }

    private void SetMonstersBeaten()
    {
        foreach (GameObject m in GameObject.FindGameObjectsWithTag("Monster"))
        {
            m.GetComponent<MonsterController>().SetBeaten(true);
        }
    }
}
